src/drf_yasg/inspectors/field.py

```diff
@@ -455,18 +455,89 @@ def decimal_return_type():
 
 
-raw_type_info = [
+hinting_type_info = [
     (bool, (openapi.TYPE_BOOLEAN, None)),
     (int, (openapi.TYPE_INTEGER, None)),
+    (str, (openapi.TYPE_STRING, None)),
     (float, (openapi.TYPE_NUMBER, None)),
+    (dict, (openapi.TYPE_OBJECT, None)),
     (Decimal, (decimal_return_type, openapi.FORMAT_DECIMAL)),
     (uuid.UUID, (openapi.TYPE_STRING, openapi.FORMAT_UUID)),
     (datetime.datetime, (openapi.TYPE_STRING, openapi.FORMAT_DATETIME)),
     (datetime.date, (openapi.TYPE_STRING, openapi.FORMAT_DATE)),
-    # TODO - support typing.List etc
 ]
 
-hinting_type_info = raw_type_info
+
+def get_origin_type(hint_class):
+    origin_type = hint_class.__origin__ if hasattr(hint_class, '__origin__') else None
+    return origin_type or hint_class
+
+
+def is_origin_type_subclasses(hint_class, check_class):
+    origin_type = get_origin_type(hint_class)
+    if not inspect.isclass(origin_type):
+        return False
+    return issubclass(origin_type, check_class)
+
+
+def inspect_union_hint_class(hint_class):
+    if not typing:
+        return NotHandled
+
+    if get_origin_type(hint_class) == typing.Union:
+        child_type = hint_class.__args__[0]
+        return get_basic_type_info_from_hint(child_type)
+
+    return NotHandled
+
+
+def inspect_primitive_hint_class(hint_class):
+    for check_class, type_format in hinting_type_info:
+        if is_origin_type_subclasses(hint_class, check_class):
+            swagger_type, format = type_format
+            if callable(swagger_type):
+                swagger_type = swagger_type()
+            return OrderedDict([
+                ('type', swagger_type),
+                ('format', format),
+            ])
+
+    return NotHandled
+
+
+def inspect_collection_hint_class(hint_class):
+    if not typing:
+        return NotHandled
+
+    if is_origin_type_subclasses(hint_class, (typing.Sequence, typing.AbstractSet)):
+        args = hint_class.__args__
+        child_class = args[0] if args else str
+        child_type_info = get_basic_type_info_from_hint(child_class)
+        if not child_type_info:
+            child_type_info = {'type': openapi.TYPE_STRING}
+        return OrderedDict([
+            ('type', openapi.TYPE_ARRAY),
+            ('items', openapi.Items(**child_type_info)),
+        ])
+
+    return NotHandled
+
+
+hint_class_inspectors = [
+    inspect_union_hint_class,
+    inspect_primitive_hint_class,
+    inspect_collection_hint_class,
+]
+
+
+def register_hint_class_inspector(inspector):
+    """Registers a user defined type hint inspector, which converts it to a swagger type.
+    It can be used for a custom JSON serializable types, like enums.
+
+    :param inspector: a callable of one argument of type hint class.
+    It should return a swagger type keywords arguments dict or NotHandled if the type wasn't handled.
+    """
+    hint_class_inspectors.append(inspector)
```

Comment:
I don't quite like this. I don't think we should start providing an interface for type hint inspection, I don't see it as something important enough. 

Users can append to the global array themselves, if they so wish, but do so while acknowledging that it's not an explicitly supported extension point.